### PR TITLE
update ModelSummary

### LINKED ISSUE #20944: ModelSummary can't show the mode of model whitch is frozen
### Description & Motivation

Current ModelSummary model can only show the mode of module in two state, either train or eval, but we want to know if the module is frozen.

### Pitch

_No response_

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by YChienHung**
## What does this PR do?

Fixes #20944 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--20945.org.readthedocs.build/en/20945/

### CODE REVIEW

File: src/lightning/pytorch/utilities/model_summary/model_summary.py [hunk 1]
```diff
@@ -145,6 +145,13 @@ def training(self) -> bool:
         """Returns whether the module is in training mode."""
         return self._module.training
 
+    @property
+    def requires_grad(self) -> bool:
+        """Returns whether the module is requires grad."""
+        if self.num_parameters > 0:
+            return any(param.requires_grad for name, param in self._module.named_parameters())
+        return True
```
Comment (Borda) [inline src/lightning/pytorch/utilities/model_summary/model_summary.py:161]:
```suggestion
        return False
```
It may require grad if there are no parameters?

Comment (YChienHung) [reply src/lightning/pytorch/utilities/model_summary/model_summary.py:161]: Yes, my strategy is the module with no paramters is set to require grad to make sure that its mode is either train or eval. So it should return True.

**Comment by YChienHung** [comment]
@Borda Hi,Borda, is there a problem with my design ?

**Review by Borda — APPROVED** [review]

File: src/lightning/pytorch/utilities/model_summary/model_summary.py [hunk 2]
```diff
@@ -265,8 +272,8 @@ def param_nums(self) -> list[int]:
         return [layer.num_parameters for layer in self._layer_summary.values()]
 
     @property
-    def training_modes(self) -> list[bool]:
-        return [layer.training for layer in self._layer_summary.values()]
+    def training_modes(self) -> list[int]:
+        return [(2 if layer.training else 1) if layer.requires_grad else 0 for layer in self._layer_summary.values()]
```
Comment (Borda) [inline src/lightning/pytorch/utilities/model_summary/model_summary.py]: lets rather use enum

**Review by Borda — CHANGES_REQUESTED** [review]
Let's first clarify what the desired state is https://github.com/Lightning-AI/pytorch-lightning/issues/20944#issuecomment-3168875837

File: src/lightning/pytorch/utilities/model_summary/model_summary.py (hunk 2 above)
Comment (YChienHung) [reply src/lightning/pytorch/utilities/model_summary/model_summary.py]: That's a good idea !

File: src/lightning/pytorch/utilities/model_summary/model_summary.py [hunk 3]
```diff
@@ -281,14 +295,32 @@ def param_nums(self) -> list[int]:
...
+        #         if any(p.requires_grad for p in layer.parameters())
+        #         else ModelSummaryTrainingMode.FREEZE
+        #     ).value
+        #     for layer in islice(self._model.modules(), 1, None)  # exclude the root module
+        # ]
+        # return {
+        #     "train": modes.count(ModelSummaryTrainingMode.TRAIN.value),
+        #     "eval": modes.count(ModelSummaryTrainingMode.EVAL.value),
+        #     "freeze": modes.count(ModelSummaryTrainingMode.FREEZE.value),
+        # }
```
Comment (Borda) [inline src/lightning/pytorch/utilities/model_summary/model_summary.py:323]:
```suggestion
```
is it needed?
